connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala

```diff
@@ -152,6 +156,26 @@ case class SessionHolder(userId: String, sessionId: String, session: SparkSessio
   private[connect] def removeCachedDataFrame(dfId: String): DataFrame = {
     dataFrameCache.remove(dfId)
   }
+
+  /**
+   * Caches given StreamingQueryListener with the ID.
+   */
+  private[connect] def cacheListenerById(id: String, listener: StreamingQueryListener): Unit = {
+    if (listenerCache.putIfAbsent(id, listener) != null) {
+      SparkException.internalError(s"A listener is already associated with id $id")
+    }
+  }
+
+  /**
+   * Returns [[StreamingQueryListener]] cached for Listener ID `id`. If it is not found, throw
+   * [[InvalidPlanInput]].
+   */
+  private[connect] def getListenerOrThrow(id: String): StreamingQueryListener = {
```

Comment:
There does not seem to be 'remove()' method. When are these entries removed?

Reply:
Added remove() method which would be called in `removeListener()`